{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "08a7ba63",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from cairn import pca\n",
    "#from cairn.io import load_genotype_array\n",
    "from yaspin import yaspin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "36204d4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>callset_order</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>conc (ng/ul)</th>\n",
       "      <th>sequencing_date</th>\n",
       "      <th>sampleset_id</th>\n",
       "      <th>fastq_1_path</th>\n",
       "      <th>fastq_2_path</th>\n",
       "      <th>bam_path</th>\n",
       "      <th>vcf_path</th>\n",
       "      <th>legacy_id</th>\n",
       "      <th>...</th>\n",
       "      <th>longitude</th>\n",
       "      <th>sequencing_qc_pass</th>\n",
       "      <th>data_qc_pass</th>\n",
       "      <th>analysis_pop</th>\n",
       "      <th>pop_code</th>\n",
       "      <th>SuperPopulation</th>\n",
       "      <th>collection_method</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>note</th>\n",
       "      <th>Unnamed: 23</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>A01</td>\n",
       "      <td>5.02</td>\n",
       "      <td>01/07/2025</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.k...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>40.158987</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ETN</td>\n",
       "      <td>Horn of Africa &amp; Yemen</td>\n",
       "      <td>Prokopack</td>\n",
       "      <td>01/11/2024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>A02</td>\n",
       "      <td>4.28</td>\n",
       "      <td>01/07/2025</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.k...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>40.158987</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ETN</td>\n",
       "      <td>Horn of Africa &amp; Yemen</td>\n",
       "      <td>Prokopack</td>\n",
       "      <td>01/11/2024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>A03</td>\n",
       "      <td>3.60</td>\n",
       "      <td>01/07/2025</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.k...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>40.158987</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ETN</td>\n",
       "      <td>Horn of Africa &amp; Yemen</td>\n",
       "      <td>Prokopack</td>\n",
       "      <td>01/11/2024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>A04</td>\n",
       "      <td>6.92</td>\n",
       "      <td>01/07/2025</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.k...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>40.158987</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ETN</td>\n",
       "      <td>Horn of Africa &amp; Yemen</td>\n",
       "      <td>Prokopack</td>\n",
       "      <td>01/11/2024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>A05</td>\n",
       "      <td>18.00</td>\n",
       "      <td>01/07/2025</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.k...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>40.158987</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ETN</td>\n",
       "      <td>Horn of Africa &amp; Yemen</td>\n",
       "      <td>Prokopack</td>\n",
       "      <td>01/11/2024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>596</td>\n",
       "      <td>Yeme_A_13</td>\n",
       "      <td>12.53</td>\n",
       "      <td>28/04/2024</td>\n",
       "      <td>eth.djib.yem.20240606</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>Yeme_A_12_EKDN240018904-1A_22JVYWLT3_L7</td>\n",
       "      <td>...</td>\n",
       "      <td>45.020000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yemen</td>\n",
       "      <td>YEM</td>\n",
       "      <td>Horn of Africa &amp; Yemen</td>\n",
       "      <td>CDC</td>\n",
       "      <td>01/07/2022</td>\n",
       "      <td>collected btw 07/2022-02/2023</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>597</th>\n",
       "      <td>597</td>\n",
       "      <td>Yeme_A_14</td>\n",
       "      <td>20.47</td>\n",
       "      <td>28/04/2024</td>\n",
       "      <td>eth.djib.yem.20240606</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>Yeme_A_13_EKDN240018905-1A_223WLHLT4_L2</td>\n",
       "      <td>...</td>\n",
       "      <td>45.020000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yemen</td>\n",
       "      <td>YEM</td>\n",
       "      <td>Horn of Africa &amp; Yemen</td>\n",
       "      <td>CDC</td>\n",
       "      <td>01/07/2022</td>\n",
       "      <td>collected btw 07/2022-02/2023</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>598</td>\n",
       "      <td>Yeme_A_15</td>\n",
       "      <td>14.93</td>\n",
       "      <td>28/04/2024</td>\n",
       "      <td>eth.djib.yem.20240606</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>Yeme_A_14_EKDN240018906-1A_22JVYWLT3_L7</td>\n",
       "      <td>...</td>\n",
       "      <td>45.020000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yemen</td>\n",
       "      <td>YEM</td>\n",
       "      <td>Horn of Africa &amp; Yemen</td>\n",
       "      <td>CDC</td>\n",
       "      <td>01/07/2022</td>\n",
       "      <td>collected btw 07/2022-02/2023</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>599</td>\n",
       "      <td>Yeme_A_17</td>\n",
       "      <td>19.29</td>\n",
       "      <td>28/04/2024</td>\n",
       "      <td>eth.djib.yem.20240606</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>Yeme_A_15_EKDN240018907-1A_22772HLT4_L8</td>\n",
       "      <td>...</td>\n",
       "      <td>45.020000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yemen</td>\n",
       "      <td>YEM</td>\n",
       "      <td>Horn of Africa &amp; Yemen</td>\n",
       "      <td>CDC</td>\n",
       "      <td>01/07/2022</td>\n",
       "      <td>collected btw 07/2022-02/2023</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>600</td>\n",
       "      <td>Yeme_A_19</td>\n",
       "      <td>4.87</td>\n",
       "      <td>28/04/2024</td>\n",
       "      <td>eth.djib.yem.20240606</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>/home/dennist/lstm_data/cease/reads_gvcf/eth.y...</td>\n",
       "      <td>Yeme_A_17_EKDN240018909-1A_22JVYWLT3_L7</td>\n",
       "      <td>...</td>\n",
       "      <td>45.020000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yemen</td>\n",
       "      <td>YEM</td>\n",
       "      <td>Horn of Africa &amp; Yemen</td>\n",
       "      <td>CDC</td>\n",
       "      <td>01/07/2022</td>\n",
       "      <td>collected btw 07/2022-02/2023</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>601 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     callset_order  sample_id  conc (ng/ul) sequencing_date  \\\n",
       "0                0        A01          5.02      01/07/2025   \n",
       "1                1        A02          4.28      01/07/2025   \n",
       "2                2        A03          3.60      01/07/2025   \n",
       "3                3        A04          6.92      01/07/2025   \n",
       "4                4        A05         18.00      01/07/2025   \n",
       "..             ...        ...           ...             ...   \n",
       "596            596  Yeme_A_13         12.53      28/04/2024   \n",
       "597            597  Yeme_A_14         20.47      28/04/2024   \n",
       "598            598  Yeme_A_15         14.93      28/04/2024   \n",
       "599            599  Yeme_A_17         19.29      28/04/2024   \n",
       "600            600  Yeme_A_19          4.87      28/04/2024   \n",
       "\n",
       "              sampleset_id                                       fastq_1_path  \\\n",
       "0                      NaN                                                NaN   \n",
       "1                      NaN                                                NaN   \n",
       "2                      NaN                                                NaN   \n",
       "3                      NaN                                                NaN   \n",
       "4                      NaN                                                NaN   \n",
       "..                     ...                                                ...   \n",
       "596  eth.djib.yem.20240606  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "597  eth.djib.yem.20240606  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "598  eth.djib.yem.20240606  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "599  eth.djib.yem.20240606  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "600  eth.djib.yem.20240606  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "\n",
       "                                          fastq_2_path  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "..                                                 ...   \n",
       "596  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "597  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "598  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "599  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "600  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "\n",
       "                                              bam_path  \\\n",
       "0    /home/dennist/lstm_data/cease/reads_gvcf/eth.k...   \n",
       "1    /home/dennist/lstm_data/cease/reads_gvcf/eth.k...   \n",
       "2    /home/dennist/lstm_data/cease/reads_gvcf/eth.k...   \n",
       "3    /home/dennist/lstm_data/cease/reads_gvcf/eth.k...   \n",
       "4    /home/dennist/lstm_data/cease/reads_gvcf/eth.k...   \n",
       "..                                                 ...   \n",
       "596  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "597  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "598  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "599  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "600  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "\n",
       "                                              vcf_path  \\\n",
       "0                                                  NaN   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "..                                                 ...   \n",
       "596  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "597  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "598  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "599  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "600  /home/dennist/lstm_data/cease/reads_gvcf/eth.y...   \n",
       "\n",
       "                                   legacy_id  ...  longitude  \\\n",
       "0                                        NaN  ...  40.158987   \n",
       "1                                        NaN  ...  40.158987   \n",
       "2                                        NaN  ...  40.158987   \n",
       "3                                        NaN  ...  40.158987   \n",
       "4                                        NaN  ...  40.158987   \n",
       "..                                       ...  ...        ...   \n",
       "596  Yeme_A_12_EKDN240018904-1A_22JVYWLT3_L7  ...  45.020000   \n",
       "597  Yeme_A_13_EKDN240018905-1A_223WLHLT4_L2  ...  45.020000   \n",
       "598  Yeme_A_14_EKDN240018906-1A_22JVYWLT3_L7  ...  45.020000   \n",
       "599  Yeme_A_15_EKDN240018907-1A_22772HLT4_L8  ...  45.020000   \n",
       "600  Yeme_A_17_EKDN240018909-1A_22JVYWLT3_L7  ...  45.020000   \n",
       "\n",
       "    sequencing_qc_pass data_qc_pass  analysis_pop  pop_code  \\\n",
       "0                  NaN          NaN           NaN       ETN   \n",
       "1                  NaN          NaN           NaN       ETN   \n",
       "2                  NaN          NaN           NaN       ETN   \n",
       "3                  NaN          NaN           NaN       ETN   \n",
       "4                  NaN          NaN           NaN       ETN   \n",
       "..                 ...          ...           ...       ...   \n",
       "596                1.0          1.0         Yemen       YEM   \n",
       "597                1.0          1.0         Yemen       YEM   \n",
       "598                1.0          1.0         Yemen       YEM   \n",
       "599                1.0          1.0         Yemen       YEM   \n",
       "600                1.0          1.0         Yemen       YEM   \n",
       "\n",
       "            SuperPopulation  collection_method collection_date  \\\n",
       "0    Horn of Africa & Yemen          Prokopack      01/11/2024   \n",
       "1    Horn of Africa & Yemen          Prokopack      01/11/2024   \n",
       "2    Horn of Africa & Yemen          Prokopack      01/11/2024   \n",
       "3    Horn of Africa & Yemen          Prokopack      01/11/2024   \n",
       "4    Horn of Africa & Yemen          Prokopack      01/11/2024   \n",
       "..                      ...                ...             ...   \n",
       "596  Horn of Africa & Yemen                CDC      01/07/2022   \n",
       "597  Horn of Africa & Yemen                CDC      01/07/2022   \n",
       "598  Horn of Africa & Yemen                CDC      01/07/2022   \n",
       "599  Horn of Africa & Yemen                CDC      01/07/2022   \n",
       "600  Horn of Africa & Yemen                CDC      01/07/2022   \n",
       "\n",
       "                              note Unnamed: 23  \n",
       "0                              NaN         NaN  \n",
       "1                              NaN         NaN  \n",
       "2                              NaN         NaN  \n",
       "3                              NaN         NaN  \n",
       "4                              NaN         NaN  \n",
       "..                             ...         ...  \n",
       "596  collected btw 07/2022-02/2023         NaN  \n",
       "597  collected btw 07/2022-02/2023         NaN  \n",
       "598  collected btw 07/2022-02/2023         NaN  \n",
       "599  collected btw 07/2022-02/2023         NaN  \n",
       "600  collected btw 07/2022-02/2023         NaN  \n",
       "\n",
       "[601 rows x 24 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define zarr base path and metadata\n",
    "zarr_p = \"/Users/dennistpw/Projects/AsGARD/data/variants_combined_cohorts/combined_cohorts.{contig}.zarr\"\n",
    "metadata_p =  \"/Users/dennistpw/Projects/AsGARD/metadata/cease_combinedmetadata.20250715.tsv\"\n",
    "\n",
    "\n",
    "metadata = pd.read_csv(metadata_p, sep='\\t')\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "48a51ec9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False,  True,  True,  True,  True,  True,  True, False,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True,  True,  True,  True,  True,\n",
       "        True,  True,  True,  True,  True, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "bool_query = np.array(metadata.eval('country == \"Ethiopia\"'))\n",
    "bool_query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f8278051",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running analysis: cce542dac2f0dee394660f9fdc87992d\n",
      "⠏                      "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/dennistpw/miniconda3/envs/cairn/lib/python3.11/site-packages/allel/stats/preprocessing.py:134: RuntimeWarning: invalid value encountered in divide\n",
      "  gn /= self.std_\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   \r"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "array must not contain infs or NaNs",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mValueError\u001b[39m                                Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[4]\u001b[39m\u001b[32m, line 16\u001b[39m\n\u001b[32m      1\u001b[39m     \u001b[38;5;66;03m# zarr_base_path: str,\u001b[39;00m\n\u001b[32m      2\u001b[39m     \u001b[38;5;66;03m# region: str,\u001b[39;00m\n\u001b[32m      3\u001b[39m     \u001b[38;5;66;03m# df_samples: pd.DataFrame,\u001b[39;00m\n\u001b[32m   (...)\u001b[39m\u001b[32m     12\u001b[39m     \u001b[38;5;66;03m# filter_mask: str = None,\u001b[39;00m\n\u001b[32m     13\u001b[39m     \u001b[38;5;66;03m# analysis_name: str = None,\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m16\u001b[39m p = \u001b[43mpca\u001b[49m\u001b[43m.\u001b[49m\u001b[43mrun_pca\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m     17\u001b[39m \u001b[43m    \u001b[49m\u001b[43mzarr_base_path\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[43mzarr_p\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     18\u001b[39m \u001b[43m    \u001b[49m\u001b[43mregion\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mCM023248:1000000-20000000\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     19\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdf_samples\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[43mmetadata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m     20\u001b[39m \u001b[43m    \u001b[49m\u001b[43mn_snps\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m50_000\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     21\u001b[39m \u001b[43m    \u001b[49m\u001b[43mgenotype_var\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mcall_genotype\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     22\u001b[39m \u001b[43m    \u001b[49m\u001b[43mpos_var\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mvariant_position\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     23\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcache_path\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43m./results_cache/\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     24\u001b[39m \u001b[43m    \u001b[49m\u001b[43msample_query\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mcountry == \u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mEthiopia\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m'\u001b[39;49m\n\u001b[32m     25\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/github/tristanpwdennis/cairn/src/cairn/pca.py:158\u001b[39m, in \u001b[36mrun_pca\u001b[39m\u001b[34m(zarr_base_path, region, df_samples, genotype_var, pos_var, cache_path, n_components, min_minor_ac, thin_offset, n_snps, sample_query, filter_mask, analysis_name)\u001b[39m\n\u001b[32m    156\u001b[39m \u001b[38;5;66;03m# Run PCA\u001b[39;00m\n\u001b[32m    157\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m yaspin(\u001b[33m\"\u001b[39m\u001b[33mComputing PCA...\u001b[39m\u001b[33m\"\u001b[39m):\n\u001b[32m--> \u001b[39m\u001b[32m158\u001b[39m     coords, model = \u001b[43mallel\u001b[49m\u001b[43m.\u001b[49m\u001b[43mpca\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgn\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn_components\u001b[49m\u001b[43m=\u001b[49m\u001b[43mn_components\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mscaler\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mpatterson\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m \u001b[38;5;66;03m# Run PCA\u001b[39;00m\n\u001b[32m    159\u001b[39m     df_coords = pd.DataFrame(\n\u001b[32m    160\u001b[39m         {\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mPC\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mi\u001b[38;5;250m \u001b[39m+\u001b[38;5;250m \u001b[39m\u001b[32m1\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m: coords[:, i] \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(coords.shape[\u001b[32m1\u001b[39m])}\n\u001b[32m    161\u001b[39m     )    \n\u001b[32m    162\u001b[39m     data = pd.concat([df.reset_index(), df_coords.reset_index(drop=\u001b[38;5;28;01mTrue\u001b[39;00m)], axis=\u001b[32m1\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/cairn/lib/python3.11/site-packages/allel/stats/decomposition.py:55\u001b[39m, in \u001b[36mpca\u001b[39m\u001b[34m(gn, n_components, copy, scaler, ploidy)\u001b[39m\n\u001b[32m     52\u001b[39m model = GenotypePCA(n_components, copy=copy, scaler=scaler, ploidy=ploidy)\n\u001b[32m     54\u001b[39m \u001b[38;5;66;03m# fit the model and project the input data onto the new dimensions\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m55\u001b[39m coords = \u001b[43mmodel\u001b[49m\u001b[43m.\u001b[49m\u001b[43mfit_transform\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     57\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m coords, model\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/cairn/lib/python3.11/site-packages/allel/stats/decomposition.py:74\u001b[39m, in \u001b[36mGenotypePCA.fit_transform\u001b[39m\u001b[34m(self, gn)\u001b[39m\n\u001b[32m     73\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mfit_transform\u001b[39m(\u001b[38;5;28mself\u001b[39m, gn):\n\u001b[32m---> \u001b[39m\u001b[32m74\u001b[39m     u, s, v = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_fit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     75\u001b[39m     u = u[:, :\u001b[38;5;28mself\u001b[39m.n_components]\n\u001b[32m     76\u001b[39m     u *= s[:\u001b[38;5;28mself\u001b[39m.n_components]\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/cairn/lib/python3.11/site-packages/allel/stats/decomposition.py:91\u001b[39m, in \u001b[36mGenotypePCA._fit\u001b[39m\u001b[34m(self, gn)\u001b[39m\n\u001b[32m     88\u001b[39m n_samples, n_features = x.shape\n\u001b[32m     90\u001b[39m \u001b[38;5;66;03m# singular value decomposition\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m91\u001b[39m u, s, v = \u001b[43mscipy\u001b[49m\u001b[43m.\u001b[49m\u001b[43mlinalg\u001b[49m\u001b[43m.\u001b[49m\u001b[43msvd\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfull_matrices\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[32m     93\u001b[39m \u001b[38;5;66;03m# calculate explained variance\u001b[39;00m\n\u001b[32m     94\u001b[39m explained_variance_ = (s ** \u001b[32m2\u001b[39m) / n_samples\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/cairn/lib/python3.11/site-packages/scipy/_lib/_util.py:1233\u001b[39m, in \u001b[36m_apply_over_batch.<locals>.decorator.<locals>.wrapper\u001b[39m\u001b[34m(*args, **kwargs)\u001b[39m\n\u001b[32m   1231\u001b[39m \u001b[38;5;66;03m# Early exit if call is not batched\u001b[39;00m\n\u001b[32m   1232\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28many\u001b[39m(batch_shapes):\n\u001b[32m-> \u001b[39m\u001b[32m1233\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mf\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43marrays\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43mother_args\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1235\u001b[39m \u001b[38;5;66;03m# Determine broadcasted batch shape\u001b[39;00m\n\u001b[32m   1236\u001b[39m batch_shape = np.broadcast_shapes(*batch_shapes)  \u001b[38;5;66;03m# Gives OK error message\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/cairn/lib/python3.11/site-packages/scipy/linalg/_decomp_svd.py:110\u001b[39m, in \u001b[36msvd\u001b[39m\u001b[34m(a, full_matrices, compute_uv, overwrite_a, check_finite, lapack_driver)\u001b[39m\n\u001b[32m     16\u001b[39m \u001b[38;5;129m@_apply_over_batch\u001b[39m((\u001b[33m'\u001b[39m\u001b[33ma\u001b[39m\u001b[33m'\u001b[39m, \u001b[32m2\u001b[39m))\n\u001b[32m     17\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34msvd\u001b[39m(a, full_matrices=\u001b[38;5;28;01mTrue\u001b[39;00m, compute_uv=\u001b[38;5;28;01mTrue\u001b[39;00m, overwrite_a=\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[32m     18\u001b[39m         check_finite=\u001b[38;5;28;01mTrue\u001b[39;00m, lapack_driver=\u001b[33m'\u001b[39m\u001b[33mgesdd\u001b[39m\u001b[33m'\u001b[39m):\n\u001b[32m     19\u001b[39m \u001b[38;5;250m    \u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m     20\u001b[39m \u001b[33;03m    Singular Value Decomposition.\u001b[39;00m\n\u001b[32m     21\u001b[39m \n\u001b[32m   (...)\u001b[39m\u001b[32m    108\u001b[39m \n\u001b[32m    109\u001b[39m \u001b[33;03m    \"\"\"\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m110\u001b[39m     a1 = \u001b[43m_asarray_validated\u001b[49m\u001b[43m(\u001b[49m\u001b[43ma\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck_finite\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcheck_finite\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    111\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(a1.shape) != \u001b[32m2\u001b[39m:\n\u001b[32m    112\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[33m'\u001b[39m\u001b[33mexpected matrix\u001b[39m\u001b[33m'\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/cairn/lib/python3.11/site-packages/scipy/_lib/_util.py:455\u001b[39m, in \u001b[36m_asarray_validated\u001b[39m\u001b[34m(a, check_finite, sparse_ok, objects_ok, mask_ok, as_inexact)\u001b[39m\n\u001b[32m    453\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[33m'\u001b[39m\u001b[33mmasked arrays are not supported\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m    454\u001b[39m toarray = np.asarray_chkfinite \u001b[38;5;28;01mif\u001b[39;00m check_finite \u001b[38;5;28;01melse\u001b[39;00m np.asarray\n\u001b[32m--> \u001b[39m\u001b[32m455\u001b[39m a = \u001b[43mtoarray\u001b[49m\u001b[43m(\u001b[49m\u001b[43ma\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    456\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m objects_ok:\n\u001b[32m    457\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m a.dtype \u001b[38;5;129;01mis\u001b[39;00m np.dtype(\u001b[33m'\u001b[39m\u001b[33mO\u001b[39m\u001b[33m'\u001b[39m):\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/miniconda3/envs/cairn/lib/python3.11/site-packages/numpy/lib/_function_base_impl.py:665\u001b[39m, in \u001b[36masarray_chkfinite\u001b[39m\u001b[34m(a, dtype, order)\u001b[39m\n\u001b[32m    663\u001b[39m a = asarray(a, dtype=dtype, order=order)\n\u001b[32m    664\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m a.dtype.char \u001b[38;5;129;01min\u001b[39;00m typecodes[\u001b[33m'\u001b[39m\u001b[33mAllFloat\u001b[39m\u001b[33m'\u001b[39m] \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m np.isfinite(a).all():\n\u001b[32m--> \u001b[39m\u001b[32m665\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[32m    666\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33marray must not contain infs or NaNs\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    667\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m a\n",
      "\u001b[31mValueError\u001b[39m: array must not contain infs or NaNs"
     ]
    }
   ],
   "source": [
    "    # zarr_base_path: str,\n",
    "    # region: str,\n",
    "    # df_samples: pd.DataFrame,\n",
    "    # genotype_var: str =\"calldata/GT\",\n",
    "    # pos_var: str = \"variants/POS\",\n",
    "    # cache_path : str = os.getcwd(),\n",
    "    # n_components : int = 10,\n",
    "    # min_minor_ac: Union[int, float] = 1,\n",
    "    # thin_offset: int = 0,\n",
    "    # n_snps: int = None,\n",
    "    # sample_query: str = None,\n",
    "    # filter_mask: str = None,\n",
    "    # analysis_name: str = None,\n",
    "\n",
    "\n",
    "p = pca.run_pca(\n",
    "    zarr_base_path = zarr_p,\n",
    "    region = \"CM023248:1000000-20000000\",\n",
    "    df_samples = metadata,\n",
    "    n_snps=50_000,\n",
    "    genotype_var ='call_genotype',\n",
    "    pos_var = \"variant_position\",\n",
    "    cache_path = './results_cache/',\n",
    "    sample_query = 'country == \"Ethiopia\"'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c895a358",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "⠼ Loading genotypes... "
     ]
    }
   ],
   "source": [
    "# Prepare inputs\n",
    "from cairn.io import load_genotype_array, _select_random_elements_sorted\n",
    "with yaspin(\"Preparing input matrix...\"):\n",
    "    g = load_genotype_array(\n",
    "    zarr_base_path = zarr_p,\n",
    "    region = \"CM023248:1000000-20000000\",\n",
    "    df_samples = metadata,\n",
    "    genotype_var ='call_genotype',\n",
    "    pos_var = \"variant_position\",\n",
    "    sample_query = 'country == \"Ethiopia\"'\n",
    "    )\n",
    "\n",
    "    ac = g.count_alleles() #Count alleles\n",
    "    flt = (ac.max_allele() == 1) & (ac[:, :2].min(axis=1) > 1) & (ac.is_segregating()) # Remove singletons and multiallelics, select only segregating sites\n",
    "    gf = g.compress(flt, axis=0) #Apply filter\n",
    "    gn = _select_random_elements_sorted(gf, 50_000).to_n_alt() # Select random n snps\n",
    "    # Convert to n alt alleles\n",
    "    gn = g.to_n_alt()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e4bfab1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([False,  True,  True, ..., False, False, False], shape=(2287699,))"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sum(gn, axis=1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cairn",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
